### PR TITLE
Libimagstore/impl store create

### PR DESCRIPTION
**Pull request by matthiasbeyer**
Implementation of `Store::create`.

### CODE REVIEW

**Comment by TheNeikos** [comment]
Can you rebase so we get to see only your changes?

**Comment by matthiasbeyer** [comment]
Already done.

**Comment by matthiasbeyer** [comment]
I have to iterate over this, because of the caching part.

File: libimagstore/src/store.rs [hunk 1]
```diff
@@ -116,7 +116,58 @@ pub struct Store {
...
     pub fn create(&self, entry: Entry) -> Result<()> {
-        unimplemented!();
+        use std::fs::create_dir_all;
+        use std::io::Write;
+        use std::ops::Deref;
+
+        let dirpath = {
+            let path = entry.location();
+            if path.file_name().is_some() {
+                path.parent().unwrap().to_path_buf() // TODO: Safe here?
```
Comment (TheNeikos) [inline libimagstore/src/store.rs]: What happens if I put in 'general.conf' ?

File: libimagstore/src/store.rs [hunk 2]
```diff
@@ -116,7 +116,58 @@ pub struct Store {
...
+                file.write_all(entry.as_string().as_bytes())
+                    .map_err(|e| {
+                        StoreError::new(StoreErrorKind::IOWrite, Some(Box::new(e)))
+                    })
+                    .and_then(|_| self.cache(entry, file))
+
+            })
+    }
+
+    fn cache(&self, e: Entry, f: File) -> Result<()> {
```
Comment (TheNeikos) [inline libimagstore/src/store.rs]: Cache?

**Comment by neithernut** [comment]
For the record: targets #102.

File: libimagstore/src/store.rs (hunk 1 above)
Comment (matthiasbeyer) [inline libimagstore/src/store.rs]: You mean a non-store path?

Comment (neithernut) [inline libimagstore/src/store.rs]: I guess you don't properly deal with paths targeting files outside the store.

Comment (TheNeikos) [inline libimagstore/src/store.rs]: @neithernut That shouldn't happen

@matthiasbeyer The store itself has a path, you have to them together before doing this.

File: libimagstore/src/store.rs [hunk 3]
```diff
@@ -116,7 +116,64 @@ pub struct Store {
...
+        let id = e.location().clone();
+        let se = StoreEntry {
+            file: f,
+            entry: StoreEntryPresence::Present,
+        };
+
+        self.entries
+            .write()
+            .map_err(|e| {
+                // StoreError::new(StoreErrorKind::CacheLocking, Some(Box::new(e)))
```
Comment (TheNeikos) [inline libimagstore/src/store.rs]: Nah, you can't do that because the `WriteLockGuard` from the RwLock is not Send :)

Comment (neithernut) [inline libimagstore/src/store.rs]: Won't this entire function be removed?

Comment (TheNeikos) [inline libimagstore/src/store.rs]: I don't know, perhaps the naming is just misleading.

**Comment by matthiasbeyer** [comment]
I did some fixup and reverted the caching-functionality commits. Review pls!

**Comment by neithernut** [comment]
Ah, you introduced an IO error type. Good, because I'll need one.

File: libimagstore/src/error.rs [hunk 4]
```diff
@@ -12,13 +12,23 @@ use std::io::Error as IOError;
 pub enum StoreErrorKind {
     IdNotFound,
     OutOfMemory,
+    IO,
+    IOMkdirP,
+    IOWrite,
```
Comment (neithernut) [inline libimagstore/src/error.rs]: Wouldn't it make more sense to use the error `cause` instead of a bunch of new enum values?

**Comment by matthiasbeyer** [comment]
@neithernut feel free to `cherry-pick` the commits you need. They vanish on rebase anyways.

File: libimagstore/src/error.rs (hunk 4 above)
Comment (matthiasbeyer) [inline libimagstore/src/error.rs]: No, that's not the intension of the `cause`. With `cause` you can encapsulate errors in errors and create a "chain of errors" - which is semantically different from the underlying `cause`!

Comment (matthiasbeyer) [inline libimagstore/src/error.rs]: Its like:

* enum = What failed
* cause = Why did it fail

**Comment by neithernut** [comment]
I'll do so once my version compiles.

**Comment by matthiasbeyer** [comment]
@TheNeikos , @neithernut Review appreciated

File: libimagstore/src/store.rs [hunk 5]
```diff
@@ -125,7 +125,38 @@ impl Store {
...
+            return Err(StoreError::new(StoreErrorKind::InvalidStorePath, None));
+        }
+
+        let dirpath = {
+            if path.file_name().is_some() {
+                let mut newpath = self.location.clone();
+                newpath.push(path.parent().unwrap().to_path_buf()); // TODO: unwrap() safe here?
+                newpath
+            } else {
+                unimplemented!()
```
Comment (TheNeikos) [inline libimagstore/src/store.rs]: Do you plan on implementing something? Why not return an error?

File: libimagstore/src/entry.rs [hunk 6]
```diff
@@ -9,3 +9,15 @@ pub struct Entry {
...
 
+impl Entry {
+
+    pub fn location(&self) -> &StoreId {
+        &self.location
+    }
+
+    pub fn as_string(&self) -> String {
+        unimplemented!();
+    }
```
Comment (neithernut) [inline libimagstore/src/entry.rs]: Could you elaborate on the semantics of this method?

Comment (matthiasbeyer) [inline libimagstore/src/entry.rs]: It simply returns the content which is written to file later. So the header is parsed to text, the seperators are inserted... etc.

Comment (neithernut) [inline libimagstore/src/entry.rs]: The complete file? Header + Content? In this case, I'd prefer a different name, e.g. `dump()`.

Comment (matthiasbeyer) [inline libimagstore/src/entry.rs]: Sounds reasonable. Will change after LUG meetup.

**Comment by matthiasbeyer** [comment]
@TheNeikos ping.

File: libimagstore/src/store.rs [hunk 7]
```diff
@@ -125,7 +125,38 @@ impl Store {
...
+
+        let path = entry.get_location().clone();
+        if !path.is_absolute() {
+            return Err(StoreError::new(StoreErrorKind::InvalidStorePath, None));
+        }
+
+        let dirpath = {
+            if path.file_name().is_some() {
+                let mut newpath = self.location.clone();
+                newpath.push(path.parent().unwrap().to_path_buf()); // TODO: unwrap() safe here?
```
Comment (neithernut) [inline libimagstore/src/store.rs]: Better be paranoid.

File: libimagstore/src/entry.rs [hunk 8]
```diff
@@ -36,5 +36,9 @@ impl Entry {
         &mut self.content
     }
 
+    pub fn dump(&self) -> String {
+        unimplemented!();
+    }
+
```
Comment (neithernut) [inline libimagstore/src/entry.rs]: This might serve as a base for a point of #134.

File: libimagstore/src/store.rs (hunk 7 above)
Comment (matthiasbeyer) [inline libimagstore/src/store.rs]: The point is, if we can get a `file_name().is_some()`, we should have a `parent()` here, right? So the `unwrap()` won't fail and we wouldn't need to do complicated error handling here...

Comment (neithernut) [inline libimagstore/src/store.rs]: That's why I wrote about being paranoid. You could, at least, change it from a `TODO` to a comment explaining why you assume `unwrap()` to be save there, or omit it altogether.
Also, what if path happens to be `D:\` on a windows platform? (Not that I actually care about that platform)

**Comment by matthiasbeyer** [comment]
Done by #142 ...

**Pull request closed without merging**